src/core/server/saved_objects/migrationsv2/model.ts

```diff
@@ -418,36 +419,61 @@ export const model = (currentState: State, resW: ResponseType<AllActionStates>):
     return stateP;
   } else if (stateP.controlState === 'SET_SOURCE_WRITE_BLOCK') {
     const res = resW as ResponseType<typeof stateP.controlState>;
+    const reindexTargetMappings: IndexMapping = {
+      // @ts-expect-error we don't allow plugins to set `dynamic`
+      dynamic: false,
+      properties: {
+        type: { type: 'keyword' },
+        migrationVersion: {
+          // @ts-expect-error we don't allow plugins to set `dynamic`
+          dynamic: 'true',
+          type: 'object',
+        },
+      },
+    };
     // If the write block is successfully in place, proceed to the next step.
     if (Either.isRight(res)) {
-      stateP = { ...stateP, controlState: 'CLONE_SOURCE_TO_TARGET' };
+      stateP = {
+        ...stateP,
+        controlState: 'CREATE_REINDEX_TARGET',
+        reindexTargetMappings,
+      };
     } else if (Either.isLeft(res)) {
       if (res.left.type === 'index_not_found_exception') {
         // If the write block failed because the index doesn't exist, it means
         // another instance already completed the legacy pre-migration. Proceed
         // to the next step.
-        stateP = { ...stateP, controlState: 'CLONE_SOURCE_TO_TARGET' };
+        stateP = {
+          ...stateP,
+          controlState: 'FATAL',
+          reason: `SET_SOURCE_WRITE_BLOCK failed because the source index [${stateP.sourceIndex.value}] does not exist.`,
+        };
       }
     } else {
       return throwBadResponse(res);
     }
 
     return stateP;
-  } else if (stateP.controlState === 'CLONE_SOURCE_TO_TARGET') {
-    return { ...stateP, controlState: 'UPDATE_TARGET_MAPPINGS' };
-  } else if (stateP.controlState === 'UPDATE_TARGET_MAPPINGS') {
+  } else if (stateP.controlState === 'CREATE_REINDEX_TARGET') {
+    return { ...stateP, controlState: 'REINDEX_SOURCE_TO_TARGET' };
+  } else if (stateP.controlState === 'REINDEX_SOURCE_TO_TARGET') {
```

Comment:
One aspect that I have found confusing about reviewing this file is that we reference `stateP` here instead of the previous state. I know it's the same value, but I think it'd be easier to think about if we were checking the previous state in order to determine the next state. I also think using less math-y terms here would be good for readability: `previousState` and `nextState`